### DISCUSSION TITLE
[weakref]How to decide whether an object is alive in python?

### DISCUSSION THREAD

**Discussion by huangweiwu** [wjakob/nanobind#300] [Q&A]
In the document: bound types no longer need to be weak-referenceable, which saves a pointer per instance.
But what if I want to whether an object is alive and do something in python?
If weakref is enabled, the code below is ok. The return value of loop_func can be used to terminate the loop in C++.
Is there anyway to enable the weak-refrence.

```
o = my_ext.MyObject()
weak_o = weakref.ref(o)
def loop_func():
    wo = weak_o()
    if not wo:
        return False
    return True
my_ext.loop(loop_func)
```

**Comment by wjakob** [comment]
There is an old (never merged PR due to some issues) by @hawkinsp (see https://github.com/wjakob/nanobind/pull/100) for enabling weak references. It's not on my priority list, so external help is welcome.

**Reply by huangweiwu** [reply]
I use that PR in my test project and it seems ok? 
I subclass the test classes from within Python.
```
import gc, weakref
import my_ext

class MyStructWithWeakrefs(my_ext.StructWithWeakrefs):
    pass

class MyStructWithWeakrefsAndDynamicAttrs(my_ext.StructWithWeakrefsAndDynamicAttrs):
    pass

def test32_weak_references():
    o = MyStructWithWeakrefs(42)
    w = weakref.ref(o)
    print(w())
    del o
    gc.collect()
    print(w())

    p = MyStructWithWeakrefsAndDynamicAttrs(43)
    p.a_dynamic_attr = 101
    w = weakref.ref(p)
    assert w() is p
    assert w().a_dynamic_attr == 101
    del p
    gc.collect()
    assert w() is None

test32_weak_references()

**Reply by huangweiwu** [reply]
What do you mean by using t.Struct(101) and read the attribute using assert p.a_dynamic_attr.value() = 101?
t.Struct(101) is not equal to 101.

**Reply by wjakob** [reply]
The challenge with this PR was to get it to work on all platforms. Not sure about the test.

**Reply by huangweiwu** [reply]
The test works in CPython. Do you mean PyPy should be tested?
How to run it in PyPy environment.
In CPython, I build a my_ext.dll, rename it to my_ext.pyd and put it in the DLLs folder.

**Reply by wjakob** [reply]
How to do it on your own computer? I would say the normal way: download/install PyPy, build nanobind with it, etc.? And to be eligible for merigng, the whole test suite needs to pass on all platforms, as tested on GitHub CI.

**Reply by huangweiwu** [reply]
I'm willing to promote this work but I haven't used GitHub Cl before. And I'm not sure how to test it on all platforms.

**Reply by wjakob** [reply]
Sorry, I can't provide this level of support -- you'll have to learn how to use GitHub etc. yourself before contributing to this project.

**Reply by huangweiwu** [reply]
I've make a PR to support weakref, please check it if you have time?
